File: tutorial/main/poisson-amg.c
/*! \file poisson-amg.c
 *  \brief The first test example for FASP: using AMG to solve 
 *         the discrete Poisson equation from P1 finite element
 *
 *	\note  AMG example for FASP: C version
 *
 *  Solving the Poisson equation (P1 FEM) with AMG
 */

#include "fasp.h"
#include "fasp_functs.h"

/**
 * \fn int main (int argc, const char * argv[])
 *
 * \brief This is the main function for the first example.
 *
 * \author Chensong Zhang
 * \date   12/21/2011
 */
int main (int argc, const char * argv[]) 
{
    input_param     inparam;  // parameters from input files
    AMG_param       amgparam; // parameters for AMG
    
    printf("\n========================================");
    printf("\n||   FASP: AMG example -- C version   ||");
    printf("\n========================================\n\n");

    // Step 0. Set parameters
    // Read input and AMG parameters from a disk file
    // In this example, we read everything from a disk file:
    //          "./ini/amg.dat"
    // See the reference manual for details of the parameters. 
    fasp_param_init("ini/amg.dat",&inparam,NULL,&amgparam,NULL);
    
    // Set local parameters using the input values
    const int print_level   = inparam.print_level;
    const int problem_num   = inparam.problem_num;
    const int itsolver_type = inparam.itsolver_type;
    const int precond_type  = inparam.precond_type;
    const int output_type   = inparam.output_type;
    
    // Step 1. Get stiffness matrix and right-hand side
    // Read A and b -- P1 FE discretization for Poisson.
    // The location of the data files are given in "amg.dat".
    dCSRmat A;
    dvector b, x;
    char filename1[512], *datafile1;
    char filename2[512], *datafile2;
	
    // Read the stiffness matrix from matP1.dat
    strncpy(filename1,inparam.workdir,128);    
    datafile1="matP1.dat"; strcat(filename1,datafile1);
    fasp_dcoo_read(filename1, &A);

    // Read the RHS from rhsP1.dat
    strncpy(filename2,inparam.workdir,128);
    datafile2="rhsP1.dat"; strcat(filename2,datafile2);
    fasp_dvecind_read(filename2, &b);    
    
    // Step 2. Print problem size and AMG parameters
    if (print_level>PRINT_NONE) {
        printf("A: m = %d, n = %d, nnz = %d\n", A.row, A.col, A.nnz);
        printf("b: n = %d\n", b.row);
        fasp_param_amg_print(&amgparam);
    }
    
    // Step 3. Solve the system with AMG as an iterative solver
    // Set the initial guess to be zero and then solve it
    // with AMG method as an iterative procedure
    fasp_dvec_alloc(A.row, &x); fasp_dvec_set(A.row,&x,0.0);
    fasp_solver_amg(&A, &b, &x, &amgparam);
    
    // Step 4. Clean up memory
    fasp_dcsr_free(&A);
    fasp_dvec_free(&b);
    fasp_dvec_free(&x);
    
    return SUCCESS;
}

/*---------------------------------*/
/*--        End of File          --*/
/*---------------------------------*/

File: tutorial/main/poisson-amg.f90
!> \file poisson-amg.f90
!> \brief The first test example for FASP: using AMG to solve
!>        the discrete Poisson equation from P1 finite element
!>
!> \note  AMG example for FASP: F90 version
!>
!> Solving the Poisson equation (P1 FEM) with AMG
!>
!> \author Chensong Zhang
!> \date   12/21/2011

program test

  implicit none

  double precision, dimension(:), allocatable :: u, b, a
  integer,          dimension(:), allocatable :: ia, ja

  integer          :: iufile, n, nnz, i, prt_lvl, maxit
  double precision :: tol

  write (*,*) "=========================================="
  write (*,*) "||   FASP: AMG example -- F90 version   ||"
  write (*,*) "=========================================="
  write (*,*) ""

  ! Step 0: user defined variables
  prt_lvl = 2
  maxit = 100
  tol = 1.0d-8
  iufile = 1

  ! Step 1: read A and b 

  !===> Read data A from file
  open(unit=iufile,file='data/matFD.dat')

  read(iufile,*) n
  allocate(ia(1:n+1))   
  read(iufile,*) (ia(i),i=1,n+1)   

  nnz=ia(n+1)-ia(1)
  allocate(ja(1:nnz),a(1:nnz))
  read(iufile,*) (ja(i),i=1,nnz)
  read(iufile,*) (a(i),i=1,nnz)   

  close(iufile)

  !===> Read data b from file
  open(unit=iufile,file='data/rhsFD.dat')

  read(iufile,*) n
  allocate(b(1:n))
  read(iufile,*) (b(i),i=1,n)

  close(iufile)

  !===> Shift the index to start from 0 (for C routines)
  forall (i=1:n+1) ia(i)=ia(i)-1
  forall (i=1:nnz) ja(i)=ja(i)-1

  ! Step 2: Solve the system 

  !===> Initial guess
  allocate(u(1:n))
  u=0.0d0 
  call fasp_fwrapper_amg(n,nnz,ia,ja,a,b,u,tol,maxit,prt_lvl);
 
  ! Step 3: Clean up memory
  deallocate(ia,ja,a)
  deallocate(b,u)

end program test

!/*---------------------------------*/
!/*--        End of File          --*/
!/*---------------------------------*/